SpinW/spinw
# RELEASE PySpinW v4

Now uses the new `matlab2python` script from [libpymcr](https://github.com/pace-neutrons/libpymcr) to generate Python wrappers for Matlab code so you don't need to use `m.` and so the `pydoc` system works.

Also fixed some small bugs:

* Remove non-ascii characters in `mtimesx` which was confusing `mcc`
* Update Python `__init__.py` so initialized check is global
* Fix issue in `addmatrix` where Python could input an integer vector value breaking the code.
* Fix bug in `sw_plotspec` where handles were unassigned.
* Add a new Python script to compile `ctf` for all Matlab found on the system (using libpymcr to find the Matlab installations).
* Add wrappers for common Matlab plotting functions (`plot`, `subplot`, `[x,y]lim`, `[x,y]label`, `hold`, `gca`, `gcf` etc) allowing them to be called without the `m.` prefix.

----

To test, download the files in this zip:
[pyspinw4_checks.zip](https://github.com/SpinW/spinw/files/15500848/pyspinw4_checks.zip)

It has an IPython notebook `spinw_test.ipynb`, a CIF file needed for the notebook and a spinw wheel to install this release. Then create a new conda/mamba environment and run:

```
mamba create -n pyspinw python=3.11
mamba activate pyspinw
python -m pip install libpymcr
python -m pip install spinw-3.2.0.post32+0121850-py3-none-any.whl
python -m pip install matplotlib notebook
jupyter notebook
```

Then open the `spinw_test.ipynb` file and check that it runs without errors.

----

To compile this release do (from the root folder of the spinw repo)

```
python -m pip install libpymcr
matlab2python -a swfiles -a external --preamble "import pyspinw; m = pyspinw.Matlab()"
mv matlab_wrapped python/pyspinw/
cd python
mkdir ctf
python mcc_all.py
mkdir pyspinw/ctfs
mv ctf/SpinW*ctf pyspinw/ctfs
python -m pip install .
```

(to create a wheel use `python -m pip wheel . --no-deps` instead of the last command).
